plugins/formal-verification/formver.viper/src/org/jetbrains/kotlin/formver/viper/errors/VerificationError.kt

```diff
@@ -27,8 +20,24 @@ class VerificationError private constructor(
         }
     }
 
-    val reason: ErrorReason
-        get() = ErrorReason(result.reason())
+    /**
+     * The [offendingNode] represents the node in Viper's AST where the error occurred.
+     * This is useful with error reporting, when we want to fetch missing information from the error,
+     * and we need to reconstruct the original Kotlin's code context.
+     */
+    val offendingNode: Node
```

Comment:
This is just the naming taken from Silver.  Is there no better naming for it?

Idem for `reasonOffendingNode`.  If there is a rule to express here then that would work better than just an example.  Perhaps something like "reason's offending node is the proposition that could not be verified"?  Or are there other possibilities there?

Reply:
I am just thinking that maybe is a good idea to rename those proprieties as:

1. `offendingNode` $\to$ `errorLocationNode`: this name should be simpler. It is the AST node that points to the location where the Viper error generated.
2. `reasonOffendingNode` $\to$ `errorCauseNode`: this is the AST node pointing to the proposition that cannot be verified.


WDYT?

Reply:
Yep, that looks good!

Reply:
Actually, two nits on it:
1. We are already in an error class, `locationNode` is fine
2. `cause` and `reason` are similar, so `reasonNode` may be okay, too, as long as that matches the semantics.

Can you confirm that for ValidationErrors, `reason` is always a node expressing a condition?  Then you could make it even more specific, e.g. making `reasonNode` an `Exp` (and changing the name accordingly).

Reply:
> Can you confirm that for ValidationErrors, reason is always a node expressing a condition?

Yes, for all the test cases we have, we can assume so.

Reply:
> e.g. making `reasonNode` an `Exp`

Regarding this point here, I think that making `reasonNode` of type `Exp` requires some wrappers around the `viper.silver.ast.Exp` (taking in consideration our existing `Exp` interface). So, the solution would be something like this:

```kotlin
sealed interface Exp : IntoSilver<viper.silver.ast.Exp> {
   val type: Type
   // ...
   data class Wrapper(val exp: viper.silver.ast.Exp) : Exp {
       override val type: Type = Type.Wrapper(exp.typ())
       override fun toSilver(): viper.silver.ast.Exp = exp
   }
}

sealed interface Type : IntoSilver<viper.silver.ast.Type> {
   fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Type
   // ...
   data class Wrapper(val typ: viper.silver.ast.Type) {
       override fun substitute(typeVarMap: kotlin.collections.Map<TypeVar, Type>): Type = this
       override fun toSilver(): viper.silver.ast.Type = typ
   }

}
```

Then, I should extend the `Exp` interface with `IntoCallable` to allow the conversion of an expression as a Silver's `Callable`, similar to what happens with `Node`.

I am not a huge fan of this solution tough, WDYT? An alternative would be keeping the `reasonNode` as it is already.

Reply:
I think having a separate `WrappedExp` should work; the purpose of the class is very different than of `Exp`, so I think having both makes sense.  Having a prefix/common class they're a member of for all these wrapped types may not be a bad idea.  They're used in a very limited section of the code, so it's fine if they're more verbose than everything else.